Add support for .bss sections

jyn514 opened the pull request:
Allows defining zero-initialized data without allocating memory while writing the object file.

Closes #97.

Ongoing issues:

- [ ] [Need to test Mach-O parsing on a Mac](https://github.com/m4b/faerie/issues/97#issuecomment-533919277) (I don't own one)
- [x] Related: Need to fix [failing unit test](https://travis-ci.org/m4b/faerie/jobs/588284103#L451)/find a replacement. Additionally, tests shouldn't create random files in the current directory.
- [ ] It would be nice to emit a warning if `ZeroInit` data isn't marked as writable and allocatable, since [it will be marked that way by the ELF backend](https://github.com/jyn514/faerie/blob/bss/src/elf.rs#L323)
- [ ] Need to decide [if data sections are allowed to be `ZeroInit`](https://github.com/m4b/faerie/issues/97#issuecomment-533908225) - currently [they are not](https://github.com/jyn514/faerie/blob/bss/src/artifact.rs#L478)

philipc commented:
For Mach-O, you probably need to set `S_ZEROFILL`.

philipc commented:
For Mach-O, the `filesize` in the segment load command must not include the bss data.

> Need to test Mach-O parsing on a Mac (I don't own one), since goblin chokes on very large .bss sections

goblin should be able to handle this; it works for a test file I created (although I don't have a Mac either so I'm not 100% sure my test file is correct, but I think it is). Still would be good to test that a real Mac linker is happy with it too though.

m4b commented:
I can test it tonight

jyn514 commented:
Other things I seem to have missed (from https://github.com/aidansteele/osx-abi-macho-file-format-reference):

- > NoteSections that are to be filled with zeros must always be placed at the end of the segment.

jyn514 commented:
Getting a different error from goblin now, that's positive :P

```
thread 'bss' panicked at 'called `Result::unwrap()` on an `Err` value: Scroll(BadOffset(276447680))', src/libcore/result.rs:999:5
```

bjorn3 commented on src/artifact.rs:
```diff
@@ -95,11 +108,38 @@ struct InternalDecl {
+    /// Return the length of the data stored in this object.
+    ///
+    /// For `ZeroInit` variant, returns the amount of space
+    /// that would be taken up when the program is loaded into memory.
```
Please update this comment.

philipc commented:
You could partition the bss definitions into a separate `Vec` in `Mach::new`, similar to what is already done for cstrings.

jyn514 commented:
This is ready for review.

jyn514 commented on src/artifact.rs:
```diff
@@ -95,11 +108,38 @@ struct InternalDecl {
+    pub fn len(&self) -> usize {
+        match self {
+            Data::Blob(blob) => blob.len(),
+            Data::ZeroInit(_) => 0,
```
Yeah, `len` is no longer a good name for this, maybe `file_size` or something like that.

jyn514 commented on src/elf.rs:
```diff
@@ -480,19 +486,32 @@ impl<'a> Elf<'a> {
+        let section_name = match (def.data, decl) {
+            (Data::Blob(_), DefinedDecl::Function(_)) => format!(".text.{}", name),
+            (Data::ZeroInit(_), DefinedDecl::Function(_)) => {
+                unreachable!("cannot define function as zero-init")
```
All these `unreachable` errors are caught [here](https://github.com/m4b/faerie/pull/99/files#diff-f5613d5b06f5bdc85d896b3c98e51492R476). I'm trying to think of a way to encode this in the type system, maybe modify `DataDecl`? `DataDecl` seems to be set [in `declare`](https://github.com/jyn514/faerie/blob/bss/src/artifact.rs#L350), not after it's defined, though.

jyn514 commented on src/artifact.rs:
```diff
@@ -356,11 +396,27 @@ impl Artifact {
+    }
+
+    /// Defines a _previously declared_ program object with all zeros.
+    /// **NB**: If you attempt to define an import, this will return an error.
```
This is still an error I think: https://github.com/m4b/faerie/pull/99/files#diff-f5613d5b06f5bdc85d896b3c98e51492R463

jyn514 commented on src/mach.rs:
```diff
@@ -826,7 +865,11 @@ impl<'a> Mach<'a> {
+            if let Data::Blob(bytes) = cstring.data {
+                file.write_all(bytes)?;
+            } else {
+                unreachable!();
```
The problem I keep running into is that most of these types are set up during the _declaration_ of the data, not the _definition_. To encode this in the type system, we would have to create a new `cstring` struct during definition, which since we have to do it for a bunch of different types would make everything harder to understand IMO.

Maybe I'm just not understanding the current code very well.

jyn514 replied:
Although actually now that I look at it you're right, it's definitely possible to have a `ZeroInit` string: https://github.com/m4b/faerie/pull/99/files#diff-0a6142b10f08148adf63ae919efb879fR700

jyn514 replied:
Fixed the logic bug but it still has the `unreachable`.

jyn514 commented on src/elf.rs:
```diff
@@ -591,6 +610,28 @@ impl<'a> Elf<'a> {
+        );
+        self.nsections += 1;
+        shndx
+    }
```
Great call, I refactored this into a function.

jyn514 commented on src/artifact.rs:
```diff
@@ -376,22 +432,22 @@ impl Artifact {
         &mut self,
         name: T,
-        data: Vec<u8>,
+        data: Data,
```
Added, great idea!

jyn514 commented:
I addressed all the comments, but I'm still not sure how to encode that some data types can't be ZeroInit into the type system. Suggestions are welcome :)

bjorn3 commented:
No concerns

m4b commented:
Ok @jyn514 I noticed you had a concern about testing on macho, do you need me to do anything? If yes, just paste some commands you'd like to see, or general gist of what you're worried about; otherwise we will merge.

I'm going to squash the commits since some aren't passing, hope you don't mind.

Really great work, much appreciated, congrats! :)

jyn514 commented:
> you had a concern about testing on macho

I played around with it on a friend's Mac and it seemed to work fine. No specific concerns, just a general feeling that I missed something lol. I'm comfortable merging if you are.

> I'm going to squash the commits
+1

> Really great work, much appreciated, congrats! :)

Thank you! Next step is getting Cranelift to use the new APIs haha :)
